android/KMEA/app/src/main/java/com/tavultesoft/kmea/data/CloudCatalogDownloadCallback.java

```diff
@@ -0,0 +1,279 @@
+    }
+
+    // Filter out any duplicates from already-installed models, properly merging the lists.
+    for (int i = 0; i < lexicalModelsArrayList.size(); i++) {
+      LexicalModel model = lexicalModelsArrayList.get(i);
+
+      // Check for duplicates / possible updates.
+      LexicalModel match = aDataSet.lexicalModels.findMatch(model);
+
+      if (match != null) {
+        if (compareVersions(model, match) == FileUtils.VERSION_GREATER) {
+          aDataSet.lexicalModels.remove(match);
+        } else {
+          lexicalModelsArrayList.remove(model);
+          i--; // Decrement our index to reflect the removal.
+        }
+      } // else no match == no special handling.
+    }
+
+    // Add the cloud-returned lexical model info to the CloudRepository's lexical models adapter.
+    aDataSet.lexicalModels.addAll(lexicalModelsArrayList);
+
+    // Do the actual update checks.
+    for (int i = 0; i < installedData.lexicalModels.getCount(); i++) {
+      LexicalModel model = installedData.lexicalModels.getItem(i);
+
+      // Check for duplicates / possible updates.
+      LexicalModel match = aDataSet.lexicalModels.findMatch(model);
+
+      if (match != null) {
+        Bundle bundle = updateCheck(match, model);
+        if (bundle != null) {
+          updateBundles.add(bundle);
+        }
+      } // else no match == no special handling.
+    }
+
+    if (updateBundles.size() > 0) {
+      // Time for updates!
+      Log.v(TAG, "Performing keyboard and model updates for " + updateBundles.size() + " resources.");
+
+      updateHandler.onUpdateDetection(updateBundles);
+    }
+
+    // And finish.
+    aDataSet.notifyDataSetChanged(); // Edits are done - signal that.
+
+    if (executeCallbacks) {
+      querySuccess.run();
+    }
+  }
+
+  @Override
+  public void applyCloudDownloadToModel(Context aContext, Dataset aDataSet, CloudCatalogDownloadReturns aCloudResult)
+  {
+    saveDataToCache(aCloudResult);
+
+    ensureInitCloudReturn(aContext,aDataSet,aCloudResult);
+
+    processCloudReturns(aDataSet, aCloudResult,true);
+  }
+
+  @Override
+  public CloudCatalogDownloadReturns extractCloudResultFromDownloadSet(
+    CloudApiTypes.CloudDownloadSet<Dataset, CloudCatalogDownloadReturns> aDownload)
+  {
+    List<CloudApiTypes.CloudApiReturns> retrievedJSON = new ArrayList<>(aDownload.getSingleDownloads().size());
+
+
+    for (CloudApiTypes.SingleCloudDownload _d : aDownload.getSingleDownloads()) {
+      JSONParser jsonParser = new JSONParser();
+      JSONArray dataArray = null;
+      JSONObject dataObject = null;
+
+      if (_d.getDestinationFile() != null && _d.getDestinationFile().length() > 0) {
+        try {
+
+          if (_d.getType() == CloudApiTypes.JSONType.Array) {
+            dataArray = jsonParser.getJSONObjectFromFile(_d.getDestinationFile(),JSONArray.class);//(JSONArray) _o;
+          } else {
+            dataObject = jsonParser.getJSONObjectFromFile(_d.getDestinationFile(),JSONObject.class);//(JSONObject) _o;
```

Comment:
You can remove these commented parts now, right?

Reply:
Yes